File: VB/vb_DXDockingForLayoutPurposes/DockLayoutManagerExt.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Windows
Imports DevExpress.Xpf.Layout.Core
Imports DevExpress.Xpf.Docking
Imports DevExpress.Xpf.Docking.Platform

Namespace vb_DXDockingForLayoutPurposes
	'S171364
	Friend Class DockLayoutManagerExt
		Inherits DockLayoutManager

		Private layoutView As IView
		Private customizationView As IView

		Protected Overrides Function CreateLayoutView(ByVal element As IUIElement) As IView
			layoutView = New CustomLayoutView(element)
			Return layoutView
		End Function
		Protected Overrides Function CreateCustomizationView(ByVal element As IUIElement) As IView
			customizationView = New MyCustomizationView(element)
			Return customizationView
		End Function
	End Class

	Public Class MyCustomizationView
		Inherits CustomizationView
		Public Sub New(ByVal viewUIElement As IUIElement)
			MyBase.New(viewUIElement)
		End Sub

		Protected Overrides Sub RegisterListeners()
			MyBase.RegisterUIServiceListener(New LayoutViewSelectionListener())
			MyBase.RegisterUIServiceListener(New CustomizationViewUIInteractionListener())
			MyBase.RegisterUIServiceListener(New MyCustomizationDraggingListener()) 'my version
		End Sub
	End Class

	Public Class CustomLayoutView
		Inherits LayoutView
		Public Sub New(ByVal viewUIElement As IUIElement)
			MyBase.New(viewUIElement)
		End Sub

		Protected Overrides Sub RegisterListeners()
			RegisterUIServiceListener(New LayoutViewRegularDragListener())
			RegisterUIServiceListener(New LayoutViewFloatingDragListener())
			RegisterUIServiceListener(New LayoutViewReorderingListener())
			RegisterUIServiceListener(New MyDragListener()) ' my version
			RegisterUIServiceListener(New LayoutViewNonClientDraggingListener())
			RegisterUIServiceListener(New LayoutViewUIInteractionListener())
			RegisterUIServiceListener(New LayoutViewSelectionListener())
			RegisterUIServiceListener(New LayoutViewActionListener())
		End Sub
	End Class

	Public Class MyDragListener
		Inherits LayoutViewClientDraggingListener
		Public Overrides Function CanDrop(ByVal point As Point, ByVal element As ILayoutElement) As Boolean

			Dim dockLayoutElementDragInfo As New DockLayoutElementDragInfo(Me.View, point, element)
			Dim draggedItem As BaseLayoutItem = dockLayoutElementDragInfo.Item
			Dim target As BaseLayoutItem = dockLayoutElementDragInfo.Target
			If (Not DragHelper.IsItemInGroup(draggedItem)) AndAlso (Not DragHelper.IsTargetInGroup(target)) Then
				Return MyBase.CanDrop(point, element)
			End If
			Dim result As Boolean = DragHelper.AllowMoving(draggedItem, target)
			Return result

		End Function
	End Class

	Public Class MyCustomizationDraggingListener
		Inherits CustomizationViewClientDraggingListener
		Public Overrides Function CanDrop(ByVal point As Point, ByVal element As ILayoutElement) As Boolean
			Dim dockLayoutElementDragInfo As New DockLayoutElementDragInfo(MyBase.View, point, element)
			Dim draggedItem = dockLayoutElementDragInfo.Item
			Dim target = dockLayoutElementDragInfo.Target
			If (TypeOf dockLayoutElementDragInfo.DropTarget Is HiddenItemElement OrElse TypeOf dockLayoutElementDragInfo.DropTarget Is HiddenItemsListElement) AndAlso dockLayoutElementDragInfo.Item.AllowHide Then
				Dim itemType As LayoutItemType = dockLayoutElementDragInfo.Item.ItemType
                Return (LayoutItemsHelper.IsLayoutItem(dockLayoutElementDragInfo.Item) OrElse itemType = LayoutItemType.Group)
			End If
			If (Not DragHelper.IsItemInGroup(draggedItem)) AndAlso (Not DragHelper.IsTargetInGroup(target)) Then
				Return MyBase.CanDrop(point, element)
			End If
			Return DragHelper.AllowMoving(draggedItem, target)
		End Function
	End Class
End Namespace

File: CS/DXDockingForLayoutPurposes/DockLayoutManagerExt.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows;
using DevExpress.Xpf.Layout.Core;
using DevExpress.Xpf.Docking;
using DevExpress.Xpf.Docking.Platform;

namespace DXDockingForLayoutPurposes {
    //S171364
    class DockLayoutManagerExt : DockLayoutManager {

        IView layoutView;
        IView customizationView;

        protected override IView CreateLayoutView(IUIElement element) {
            layoutView = new CustomLayoutView(element);
            return layoutView;
        }
        protected override IView CreateCustomizationView(IUIElement element) {
            customizationView = new MyCustomizationView(element);
            return customizationView;
        }
    }

    public class MyCustomizationView : CustomizationView {
        public MyCustomizationView(IUIElement viewUIElement)
            : base(viewUIElement) {
        }

        protected override void RegisterListeners() {
            base.RegisterUIServiceListener(new LayoutViewSelectionListener());
            base.RegisterUIServiceListener(new CustomizationViewUIInteractionListener());
            base.RegisterUIServiceListener(new MyCustomizationDraggingListener());//my version
        }
    }

    public class CustomLayoutView : LayoutView {
        public CustomLayoutView(IUIElement viewUIElement)
            : base(viewUIElement) {
        }

        protected override void RegisterListeners() {
            RegisterUIServiceListener(new LayoutViewRegularDragListener());
            RegisterUIServiceListener(new LayoutViewFloatingDragListener());
            RegisterUIServiceListener(new LayoutViewReorderingListener());
            RegisterUIServiceListener(new MyDragListener());// my version
            RegisterUIServiceListener(new LayoutViewNonClientDraggingListener());
            RegisterUIServiceListener(new LayoutViewUIInteractionListener());
            RegisterUIServiceListener(new LayoutViewSelectionListener());
            RegisterUIServiceListener(new LayoutViewActionListener());
        }
    }

    public class MyDragListener : LayoutViewClientDraggingListener {
        public override bool CanDrop(Point point, ILayoutElement element) {

            DockLayoutElementDragInfo dockLayoutElementDragInfo = new DockLayoutElementDragInfo(this.View, point, element);
            BaseLayoutItem draggedItem = dockLayoutElementDragInfo.Item;
            BaseLayoutItem target = dockLayoutElementDragInfo.Target;
            if (!DragHelper.IsItemInGroup(draggedItem) && !DragHelper.IsTargetInGroup(target)) {
                return base.CanDrop(point, element);
            }
            bool result = DragHelper.AllowMoving(draggedItem, target);
            return result;

        }
    }

    public class MyCustomizationDraggingListener : CustomizationViewClientDraggingListener {
        public override bool CanDrop(Point point, ILayoutElement element) {
            DockLayoutElementDragInfo dockLayoutElementDragInfo = new DockLayoutElementDragInfo(base.View, point, element);
            var draggedItem = dockLayoutElementDragInfo.Item;
            var target = dockLayoutElementDragInfo.Target;
            if ((dockLayoutElementDragInfo.DropTarget is HiddenItemElement || dockLayoutElementDragInfo.DropTarget is HiddenItemsListElement) && dockLayoutElementDragInfo.Item.AllowHide) {
                LayoutItemType itemType = dockLayoutElementDragInfo.Item.ItemType;
                return (LayoutItemsHelper.IsLayoutItem(dockLayoutElementDragInfo.Item) || itemType == LayoutItemType.Group);
            }
            if (!DragHelper.IsItemInGroup(draggedItem) && !DragHelper.IsTargetInGroup(target)) {
                return base.CanDrop(point, element);
            }
            return DragHelper.AllowMoving(draggedItem, target);
        }
    }
}
